src/extensions/default/Inline3DParametersEditor/InlineParameterEditor.js

```diff
@@ -0,0 +1,213 @@
+    InlineParameterEditor.prototype._isOwnChange = null;
+
+    /** @type {boolean} True while we're syncing a code editor change into the parameter picker */
+    InlineParameterEditor.prototype._isHostChange = null;
+
+    /** @type {number} ID used to identify edits coming from this inline widget for undo batching */
+    InlineParameterEditor.prototype._origin = null;
+
+
+    /**
+     * Returns the current text range of the parameters we're attached to, or null if
+     * we've lost sync with what's in the code.
+     * @return {?{start:{line:number, ch:number}, end:{line:number, ch:number}}}
+     */
+    InlineParameterEditor.prototype.getCurrentRange = function () {
+        var pos, start, end;
+
+        pos = this._marker && this._marker.find();
+
+        start = pos && pos.from;
+        if (!start) {
+            return null;
+        }
+
+        end = pos.to;
+        if (!end) {
+            end = {line: start.line};
+        }
+
+        var line = this.hostEditor.document.getLine(start.line);
+
+        /**
+         * Returns an array of string that match the
+         * given Regex
+         * Eg : line = <a-cylinder position="-0.04 1.40 0.80"  rotation = "171.00 -0.10 "></a-cylinder>
+         * matches = ["-0.04 1.18 0.80", "227", " 0.34", "171.00 -0.10 "]
+         */
+        var matches = line.substr(start.ch).match(Inline3dParametersUtils.PARAMETERS_3D_REGEX);
+
+        // Note that end.ch is exclusive, so we don't need to add 1 before comparing to
+        // the matched length here.
+        if (matches && (end.ch === undefined || end.ch - start.ch < matches[0].length)) {
+            end.ch = start.ch + matches[0].length;
+            this._marker.clear();
+            this._marker = this.hostEditor._codeMirror.markText(start, end);
+        }
+
+        if (end.ch === undefined) {
+            // We were unable to resync the marker.
+            return null;
+        } else {
+            return {start: start, end: end};
+        }
+    };
+
+    /**
+     * When the selected text changes, update text in code editor
+     * @param {!string} parameterString
+     */
+    InlineParameterEditor.prototype._handleParametersChange = function (parameterString) {
+        var self = this;
+        if (parameterString !== this._parameters) {
+            var range = this.getCurrentRange();
+            if (!range) {
+                return;
+            }
+
+            // Don't push the change back into the host editor if it came from the host editor.
+            if (!this._isHostChange) {
+                var endPos = {
+                        line: range.start.line,
+                        ch: range.start.ch + parameterString.length
+                    };
+                this._isOwnChange = true;
+                this.hostEditor.document.batchOperation(function () {
+                    self.hostEditor.setSelection(range.start, range.end); // workaround for #2805
```

Comment:
If you're borrowing code, which is fine, please put a note at the top of the file(s) indicating the source (e.g., url) and the license under which you're using it.